### DISCUSSION TITLE
Custom endpoint for upload file

### DISCUSSION THREAD

**Discussion by TomS-** [directus/directus#9705] [Archive]
Hello, I have created an endpoint to capture form data. At the moment I'm just testing it to get it set up.
```
router.post('/file', async (req, res, next) => {
        try {
            console.log(req.body);
            
            data = {
                test: 'Got Here'
            };

            res.json( JSON.stringify(data) );
        } catch (error) {
            next(new InvalidQueryException(error.message));
        };
    });
```
And in Postman I'm doing:
![image](https://user-images.githubusercontent.com/4595662/141097077-540e7c20-9d65-48c5-ac29-500a8f0eb9cd.png)

The `console.log` is returning:
```
2021-11-10T10:19:23: {}
2021-11-10T10:19:23: 10:19:23 ✨ request completed POST 200 /fusioo/file 9ms
```

### To Reproduce

1. Create an post endpoint.
2. Post FormData. 
3. Try to access the request body.

**Comment by joselcvarela** [comment]
Hello @TomS- 
Directus API should work only with REST and Graphql, excluding the /files endpoint which supports multipart form, in order to receive huge payloads of data.

That being said it is not on our plans to include this functionality.
Although, I believe you can use forms on your custom hooks by adding the follow as middleware to your custom hook.
```js
router.post(
  "/file",
  express.urlencoded({
    extended: true,
  }),
  async (req, res, next) => {}
);
```

**Reply by TomS-** [reply]
Hi @joselcvarela in my case req.body is still empty with this. I ended up using busboy to do this.

**Reply by TomS-** [reply]
```
    router.post('/file', async (req, res, next) => {
        const busboy = new Busboy({ headers: req.headers });

        try {
            busboy.on('file', (fieldname, file, filename, encoding, mimetype) => {
                file.on('data', async data => {
                    const options = {
                        'method': 'POST',
                        'url': 'https://api.fusioo.com/v3/files',
                        'headers': {
                            'Authorization': `Bearer ${token}`
                        },
                        formData: {
                            'file': {
                                'value': data,
                                'options': { 'filename': filename, 'contentType': mimetype }
                            }
                        }
                    };
                    
                    request(options, (error, response) => {
                        if(error) console.log(error);
                        console.log(response.body);
                        req.json(response.body);
                        next();
                    });
                });
            });

            busboy.on('finish', () => next());
            req.pipe(busboy);

        } catch (error) {
            next(new InvalidQueryException(error.message));
        };
    });
 ```

This is what I've come up with but `req.json(response.body)` never happens but `console.log(response.body)` does. It also restarts the server before req.json happens. I'm using `request` in this instance as `axios` has issues sending files.

**Reply by joselcvarela** [reply]
Ohh I didn't know you wanted to upload a file.
In your example above, you were using a simple form.

Yeah for a custom upload file route you either need `busboy` or `multer`

**Comment by joselcvarela** [comment]
Solution given by @TomS-  
> ```
>     router.post('/file', async (req, res, next) => {
>         const busboy = new Busboy({ headers: req.headers });
> 
>         try {
>             busboy.on('file', (fieldname, file, filename, encoding, mimetype) => {
>                 file.on('data', async data => {
>                     const options = {
>                         'method': 'POST',
>                         'url': 'https://api.fusioo.com/v3/files',
>                         'headers': {
>                             'Authorization': `Bearer ${token}`
>                         },
>                         formData: {
>                             'file': {
>                                 'value': data,
>                                 'options': { 'filename': filename, 'contentType': mimetype }
>                             }
>                         }
>                     };
>                     
>                     request(options, (error, response) => {
>                         if(error) console.log(error);
>                         console.log(response.body);
>                         req.json(response.body);
>                         next();
>                     });
>                 });
>             });
> 
>             busboy.on('finish', () => next());
>             req.pipe(busboy);
> 
>         } catch (error) {
>             next(new InvalidQueryException(error.message));
>         };
>     });
> ```

Although it also should be possible to handle this with `multer`: https://github.com/expressjs/multer

**Reply by rijkvanzanten** [reply]
Busboy is exactly how directus does it as well 👍🏻 

https://github.com/directus/directus/blob/f0e9e4a60a8af6399ce67b7c591c5ce82c2440f8/api/src/controllers/files.ts#L34
